File: 09_HashMap/python/hashmap.py
class SimpleHashMap:
    def __init__(self, size=10):
        self.size = size
        self.table = [[] for _ in range(self.size)]

    def _hash(self, key):
        return hash(key) % self.size

    def put(self, key, value):
        hv = self._hash(key)
        for item in self.table[hv]:
            if item[0] == key:
                item[1] = value
                return
        self.table[hv].append([key, value])

    def get(self, key):
        hv = self._hash(key)
        for item in self.table[hv]:
            if item[0] == key:
                return item[1]
        return None

if __name__ == "__main__":
    mp = SimpleHashMap()
    mp.put("K1", "V1")
    assert mp.get("K1") == "V1", "Erreur : K1 devrait valoir V1"
    print("Test Python HashMap réussi !")

File: 09_HashMap/C/hashmap.c
#include <stdio.h>
#include <string.h>
#include <assert.h>

#define SIZE 10

struct Node {
    char key[20];
    char value[20];
};

struct Node hashTable[SIZE];

int hashFunction(char* key) {
    int sum = 0;
    for (int i = 0; key[i] != '\0'; i++) {
        sum += key[i];
    }
    return sum % SIZE;
}

void put(char* key, char* value) {
    int index = hashFunction(key);
    strcpy(hashTable[index].key, key);
    strcpy(hashTable[index].value, value);
}

char* get(char* key) {
    int index = hashFunction(key);
    return hashTable[index].value;
}

int main() {
    put("K1", "V1");
    assert(strcmp(get("K1"), "V1") == 0);
    printf("Test C HashMap réussi !\n");
    return 0;
}

